subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/ServiceReferencePropertyAnnotationHandler.java

```diff
@@ -59,4 +66,24 @@ public void visitPropertyValue(String propertyName, PropertyValue value, Propert
         }
         visitor.visitServiceReference((Provider<BuildService<?>>) value.call(), serviceName);
     }
+
+    @Override
+    public void validatePropertyMetadata(PropertyMetadata propertyMetadata, TypeValidationContext validationContext) {
+        Method getter = propertyMetadata.getGetterMethod();
+        ModelType propertyType = ModelType.returnType(getter);
+        List<ModelType<?>> typeVariables = Cast.uncheckedNonnullCast(propertyType.getTypeVariables());
+        if (typeVariables.size() != 1 || !BuildService.class.isAssignableFrom(typeVariables.get(0).getRawClass())) {
+            validationContext.visitPropertyProblem(problem ->
+                problem.withId(ValidationProblemId.SERVICE_REFERENCE_MUST_BE_A_BUILD_SERVICE)
+                    .forProperty(propertyMetadata.getPropertyName())
+                    .reportAs(ERROR)
+                    .withDescription(() -> String.format("has @ServiceReference annotation used on property of type '%s' which is not a build service implementation", typeVariables.get(0).getName()))
+                    .happensBecause(() -> String.format("A property annotated with @ServiceReference must be of a type that implements '%s'", BuildService.class.getName()))
+                    .addPossibleSolution(String.format("Make '%s' implement '%s'", typeVariables.get(0).getName(), BuildService.class.getName()))
+                    .addPossibleSolution(String.format("Replace the @ServiceReference annotation on '%s' with @Internal and assign a value of type '%s' explicitly", propertyMetadata.getPropertyName(), typeVariables.get(0).getName()))
```

Comment:
🤔 I wonder if `@Internal` is the best advice here.

Reply:
It is the current way we document consumption of build services: 

https://docs.gradle.org/current/userguide/build_services.html#using_a_build_service_from_a_task